idlparser/src/types/mod.rs

```diff
@@ -0,0 +1,374 @@
+use crate::{grammar::ProgramParser, lexer::Lexer};
+
+pub mod visitor;
+
+pub fn parse_idl(idl: &str) -> Result<Program, String> {
+    let lexer = Lexer::new(idl);
+    let parser = ProgramParser::new();
+    let program = parser.parse(lexer).map_err(|e| format!("{:?}", e))?;
+    Ok(program)
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Program {
+    service: Service,
+    types: Vec<Type>,
+}
+
+impl Program {
+    pub(crate) fn new(service: Service, types: Vec<Type>) -> Self {
+        Self { service, types }
+    }
+
+    pub fn service(&self) -> &Service {
+        &self.service
+    }
+
+    pub fn types(&self) -> &[Type] {
+        &self.types
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Service {
+    funcs: Vec<Func>,
+}
+
+impl Service {
+    pub(crate) fn new(funcs: Vec<Func>) -> Self {
+        Self { funcs }
+    }
+
+    pub fn funcs(&self) -> &[Func] {
+        &self.funcs
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Func {
+    name: String,
+    params: Vec<FuncParam>,
+    output: TypeDecl,
+    is_query: bool,
+}
+
+impl Func {
+    pub(crate) fn new(
+        name: String,
+        params: Vec<FuncParam>,
+        output: TypeDecl,
+        is_query: bool,
+    ) -> Self {
+        Self {
+            name,
+            params,
+            output,
+            is_query,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn params(&self) -> &[FuncParam] {
+        &self.params
+    }
+
+    pub fn output(&self) -> &TypeDecl {
+        &self.output
+    }
+
+    pub fn is_query(&self) -> bool {
+        self.is_query
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct FuncParam {
+    name: String,
+    type_decl: TypeDecl,
+}
+
+impl FuncParam {
+    pub(crate) fn new(name: String, type_decl: TypeDecl) -> Self {
+        Self { name, type_decl }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn type_decl(&self) -> &TypeDecl {
+        &self.type_decl
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Type {
+    name: String,
+    def: TypeDef,
```

Comment:
Wait, is this intentional? Are we forbidding top level types to be anything but structs and enums? What about writing something like 
```
type ID = u32;
type Ids = Vec<ID>` 

``` 
for semantics?